packages/inferno-router/src/resolveLoaders.ts

```diff
@@ -0,0 +1,198 @@
+import { isNullOrUndef, isUndefined } from "inferno-shared";
+import { matchPath } from "./matchPath";
+import type { TLoaderData, TLoaderProps } from "./Router";
+import { Switch } from "./Switch";
+
+export function resolveLoaders(loaderEntries: TLoaderEntry[]): Promise<Record<string, TLoaderData>> {
+  const promises = loaderEntries.map((({ path, params, request, loader }) => {
+    return resolveEntry(path, params, request, loader);
+  }));
+  return Promise.all(promises).then((result) => {
+    return Object.fromEntries(result);
+  });
+}
+
+type TLoaderEntry = {
+  path: string,
+  params: Record<string, any>,
+  request: Request,
+  controller: AbortController,
+  loader: (TLoaderProps) => Promise<TLoaderEntry>,
+}
+
+// Optionally pass base param during SSR to get fully qualified request URI passed to loader in request param
+export function traverseLoaders(location: string, tree: any, base?: string, parentIsSwitch = false): TLoaderEntry[] {
+  // Make sure tree isn't null
+  if (isNullOrUndef(tree)) return [];
+
+  if (Array.isArray(tree)) {
+    let hasMatch = false;
+    const entries = tree.reduce((res, node) => {
+      if (parentIsSwitch && hasMatch) return res;
+
+      const outpArr = traverseLoaders(location, node, base, node?.type === Switch);
+      return [...res, ...outpArr];
+    }, []);
+    return entries;
+  }
+
+
+  let outp: TLoaderEntry[] = [];
+  let isRouteButNotMatch = false;
+  if (tree.props) {
+    // TODO: If we traverse a switch, only the first match should be returned
+    // TODO: Should we check if we are in Router? It is defensive and could save a bit of time, but is it worth it?
+    const { path, exact = false, strict = false, sensitive = false } = tree.props;
+    const match = matchPath(location, {
+      path,
+      exact,
+      strict,
+      sensitive,
+    });
+
+    // So we can bail out of recursion it this was a Route which didn't match
+    isRouteButNotMatch = !match;
+
+    // Add any loader on this node (but only on the VNode)
+    if (match && !tree.context && tree.props?.loader && tree.props?.path) {
+      const { params } = match;
+      const controller = new AbortController();
+      const request = createClientSideRequest(location, controller.signal, base);
+
+      outp.push({
+        path,
+        params,
+        request,
+        controller,
+        loader: tree.props.loader,
+      })
+    }
+  }
+
+  // Traverse ends here
+  if (isRouteButNotMatch) return outp;
+
+  // Traverse children
+  const entries = traverseLoaders(location, tree.children || tree.props?.children, base, tree.type === Switch);
+  return [...outp, ...entries];
+}
+
+function resolveEntry(path, params, request, loader): Promise<any> {
+  return loader({ params, request })
+    .then((res) => [path, { res }])
+    .catch((err) => [path, { err }]);
```

Comment:
so err is `Error`  that could be typed

Reply:
Both `err` and `res` comes from the implemented loader so we don't really know what it looks like. The type won't be passed to the consuming code anyway. I will look at adding typing for the helper methods used to access the data:

WILL INVESTIGATE

ref: [packages/inferno-router/src/helpers.ts](https://github.com/infernojs/inferno/pull/1621/files#diff-ca73e06743e22d5dffd9a3baa6f39f1a0c3c49be0317cf6f46f235d3c7ecce62)

Reply:
Fixed in e6499632ed